### PR TITLE
Az.Tools.Predictor refactor and performance improvement.

### PR DESCRIPTION
**Pull request by kceiw**
## Description
There are a few things going on during the refactor: telemetry, readability, and performance.

1. Replace Tuple/ValueTuple with more meaningful types that don't use 'Item1' and 'Item2'.
2. Replace Newtonsoft Json library with the built-in System.Text.Json.
3. Reformat the telemetry events. Combine the error events with the normal event. Just have a field in the event to indicate if there is an exception or not.
4. Collect the event if the http request to the server has been sent or not. The content isn't collected.
5. Transform data format and send telemetry event in a worker in the default thread pool. This can also improve the performance since that removes the burden on the thread that's handling callback from PSReadLine.
6. Validate parameter values and also catch http response error early instead of when we try to parse the response content.
7. Remove duplicate parsing on the user input.
8. Reduce string manipulation.

## Checklist

- [X] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [X] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [X] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [X] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: tools/Az.Tools.Predictor/Az.Tools.Predictor/TelemetryData.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System;
+using System.Management.Automation.Language;
+
+namespace Microsoft.Azure.PowerShell.Tools.AzPredictor
```
Comment (mirdaki) [inline tools/Az.Tools.Predictor/Az.Tools.Predictor/TelemetryData.cs:18]: Is it normal to have many different classes in the same file?

Comment (kceiw) [reply tools/Az.Tools.Predictor/Az.Tools.Predictor/TelemetryData.cs]: I remove `TelemetryData` and move the other interface/class to seperate files.

**Comment by mirdaki** [comment]
Looks good to me!

**Review by mirdaki — APPROVED** [review]

File: tools/Az.Tools.Predictor/Az.Tools.Predictor/Telemetry/AzPredictorTelemetryClient.cs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+#endif
+        }
+
+        /// <summary>
+        /// Check whether the data collection is opted in from user.
+        /// </summary>
+        /// <returns>true if allowed.</returns>
+        private static bool IsDataCollectionAllowed()
+        {
+            if (AzurePSDataCollectionProfile.Instance.EnableAzureDataCollection == true)
```
Comment (baopingz) [inline tools/Az.Tools.Predictor/Az.Tools.Predictor/Telemetry/AzPredictorTelemetryClient.cs:160]: nit: just return AzurePSDataCollectionProfile.Instance.EnableAzureDataCollection

**Review by baopingz — APPROVED** [review]
looks good.

**Pull request merged**